assets/js/SignPanel.tsx

```diff
@@ -161,58 +161,50 @@ interface SignPanelProps {
   readOnly: boolean;
 }
 
-class SignPanel extends React.Component<
-  SignPanelProps,
-  {
-    staticLine1: string;
-    staticLine2: string;
-    customChanges: boolean;
-    initialTime: number;
-    confirmingUngroup: boolean;
-  }
-> {
-  constructor(props: SignPanelProps) {
-    super(props);
-
-    this.saveStaticText = this.saveStaticText.bind(this);
-    this.handleInputLine1 = this.handleInputLine1.bind(this);
-    this.handleInputLine2 = this.handleInputLine2.bind(this);
-
-    this.state = {
-      staticLine1: props.signConfig.line1 || '',
-      staticLine2: props.signConfig.line2 || '',
-      customChanges: false,
-      initialTime: props.currentTime,
-      confirmingUngroup: false,
-    };
-  }
-
-  handleInputLine1(newText: string): void {
-    this.setState({ staticLine1: newText, customChanges: true });
+function SignPanel({
+  alerts,
+  signId,
+  modes = {
+    auto: true,
+    custom: true,
+    headway: true,
+    off: true,
+  },
+  signContent,
+  currentTime,
+  line,
+  setConfigs,
+  signConfig,
+  realtimeId,
+  signGroup,
+  ungroupSign,
+  readOnly,
+}: SignPanelProps): JSX.Element {
+  const [staticLine1, setStaticLine1] = React.useState(signConfig.line1 || '');
+  const [staticLine2, setStaticLine2] = React.useState(signConfig.line2 || '');
+  const [customChanges, setCustomChanges] = React.useState(false);
+  const [initialTime] = React.useState(currentTime);
+  const [confirmingUngroup, setConfirmingUngroup] = React.useState(false);
+
+  function handleInputLine1(newText: string): void {
```

Comment:
Shouldn't these be defined with `useCallback`? (avoids re-defining them on every render if nothing has changed)

Reply:
Mmm, it felt like a close call to me. Whether I do:

```typescript
function handleInputLine1(newText: string): void { ... }
```

or

```typescript
const handleInputLine1 = React.useCallback(() => { .... }, [...])
```

I'm creating/allocating a new function every render either way, right? (The first is the named function, the second is the anonymous arrow function literal).

I think of the main benefit of `useCallback` as referential transparency, so that when it gets passed as a prop to a child, the *child* doesn't have to re-render. But that depends on the size of the child, and in this case it was a pretty small little component, so it didn't feel like a big win. And using `React.useCallback` makes the initial render of *this* function a little slower.

In general, I don't have a great sense of where the tradeoff is, but it doesn't make sense to use `React.useCallback` as a general rule always, and when it's a toss up I prefer the simplicity of a normal definition.

Or am I missing something about the performance trade-offs here?

Reply:
> it doesn't make sense to use `React.useCallback` as a general rule always

"Use it as a general rule always" was the impression I had, though I can't point to any official documentation that says it. If that's the wrong impression, I guess you already know more than me!

Reply:
Yeah, that was my impression after going down a rabbit hole on it a while ago, but I can't reconstruct all the things that led me to that conclusion. I think [this blog post](https://kentcdodds.com/blog/usememo-and-usecallback) was one that made a big impression. I could have sworn I saw vjeux (big react guy at facebook) say something in agreement but now I can't find that.

I do recall hearing that advice for class components about always putting logic in methods, since those really will only be created once, but even then I think it was controversial.